src/Compiler/Service/ServiceParsedInputOps.fs

```diff
@@ -1644,16 +1644,22 @@ module ParsedInput =
                     | SynType.LongIdent _ when rangeContainsPos ty.Range pos -> Some CompletionContext.Type
                     | _ -> defaultTraverse ty
 
-                member _.VisitRecordDefn(_, fields, _) =
+                member _.VisitRecordDefn(_, fields, range) =
                     fields
                     |> List.tryPick (fun (SynField(idOpt = idOpt; range = fieldRange)) ->
                         match idOpt with
                         | Some id when rangeContainsPos id.idRange pos ->
                             Some(CompletionContext.RecordField(RecordContext.Declaration true))
                         | _ when rangeContainsPos fieldRange pos -> Some(CompletionContext.RecordField(RecordContext.Declaration false))
-                        | _ -> None)
+                        | _ -> None
+                    )
                     // No completions in a record outside of all fields, except in attributes, which is established earlier in VisitAttributeApplication
-                    |> Option.orElse (Some CompletionContext.Invalid)
+                    |> Option.orElseWith (fun _ ->
+                        if rangeContainsPos range pos then
```

Comment:
If this doesn't hold, we're not in a record definition, so `VisitRecordDefn` shouldn't be called in the first place. The tree walker should therefore be fixed instead.

Reply:
> If this doesn't hold, we're not in a record definition, so VisitRecordDefn shouldn't be called in the first place. The tree walker should therefore be fixed instead.

Yes, it's outside indeed. Perhaps, fixing the walker should be the way to go.

Reply:
So, it looks like this logic was written on purpose:
https://github.com/dotnet/fsharp/blob/904b641f1295af6235cad3df28530303220ae259/src/Compiler/Service/ServiceParseTreeWalk.fs#L308-L310

We can try to change it for sure, but I'm not very familiar with this part of the compiler and don't know how many things may relate on this behavior.

Reply:
I think this changing this might break some cases with unfinished code, since we've been fixing ranges in the parser, so they don't span past last token in an unfinished node. Currently this may be mitigated by this rule in some cases.